src/proofs/ProofSteps.cpp

```diff
@@ -0,0 +1,909 @@
+                             "qnt_rm_unused"));
+
+        currStep++;
+
+        notifyObservers(Step(currStep, Step::STEP,
+                             packClause(std::make_shared<Op>("not", packClause(assumptionReNamedTerm)), unusedRem),
+                             "equiv1", std::vector<int>{currStep - 1}));
+
+        currStep++;
+
+        notifyObservers(
+            Step(currStep, Step::STEP, packClause(unusedRem), "resolution", std::vector<int>{quantStep, currStep - 1}));
+
+        currStep++;
+
+        currTerm = unusedRem;
+        assumptionReNamedTerm = unusedRem;
+        quantStep = currStep - 1;
+    }
+
+    if (not instPairs.empty()) {
+
+        notifyObservers(Step(
+            currStep, Step::STEP,
+            packClause(std::make_shared<Op>(
+                "or", packClause(std::make_shared<Op>("not", packClause(assumptionReNamedTerm)),
+                                 std::make_shared<Op>(
+                                     "!", packClause(currTerm->accept(&instantiateVisitor),
+                                                     std::make_shared<Terminal>(
+                                                         ":named " + instantiationReNamedTerm->printTerm(),
+                                                         Terminal::UNDECLARED)))))),
+            "forall_inst", instPairs));
+
+        currStep++;
+
+        notifyObservers(
+            Step(currStep, Step::STEP,
+                 packClause(std::make_shared<Op>("not", packClause(assumptionReNamedTerm)), instantiationReNamedTerm),
+                 "or", std::vector<int>{currStep - 1}));
+
+        currStep++;
+
+        currTerm = currTerm->accept(&instantiateVisitor);
+
+        notifyObservers(Step(currStep, Step::STEP, packClause(instantiationReNamedTerm), "resolution",
+                             std::vector<int>{currStep - 1, quantStep}));
+
+        currStep++;
+    }
+}
+
+void StepHandler::assumptionSteps() {
+
+    auto assertionSize = originalAssertions.size();
+    for (std::size_t i = 1; i <= assertionSize; ++i) {
+        Term * potentialLet = originalAssertions[i - 1]->accept(&letLocatorVisitor);
+        while (potentialLet != nullptr) {
+            auto simplifiedLet = potentialLet->accept(&operateLetTermVisitor);
+            SimplifyVisitor simplifyLetTermVisitor(simplifiedLet, potentialLet);
+            originalAssertions[i - 1] = originalAssertions[i - 1]->accept(&simplifyLetTermVisitor);
+            potentialLet = originalAssertions[i - 1]->accept(&letLocatorVisitor);
+        }
+        notifyObservers(Step(currStep, Step::ASSUME,
+                             packClause(std::make_shared<Op>(
+                                 "!", packClause(originalAssertions[i - 1],
+                                                 std::make_shared<Terminal>(":named @a" + std::to_string(currStep),
+                                                                            Terminal::UNDECLARED))))));
+
+        currStep++;
+    }
```

Comment:
Why do you start at 1, but then use `i-1` for access everywhere?
It seems the index can start at 0. Actually, it seems this could be turned into a range-based for loop, no?

Reply:
So, is there a reason why this cannot be a range-based for loop?

Reply:
I am changing the contents of the array for later, that is why I need the index to modify them.

Reply:
But you are only ever using the index to access the elements in `originalAssertions`.
So you can use `for (auto & assertion : originalAssertions) { ... }`

Reply:
In line 425 I am modifying the array's content. I understand the range-based for loop, I am just not sure if I can modify the contents of the array directly by changing the "auto & assertion" variable.